src/electrum_wallet_file.rs

```diff
@@ -0,0 +1,109 @@
+use crate::{ElectrumExtendedKey, ElectrumExtendedPrivKey, ElectrumExtendedPubKey};
+use regex::Regex;
+use std::fmt;
+use std::path::Path;
+use std::str::FromStr;
+
+#[derive(Debug)]
+enum WalletType {
+    Standard,
+    Multisig(u8, u8),
+}
+
+pub fn electrum_wallet_to_descriptors(json: serde_json::Value) -> Result<Vec<String>, String> {
+    let wallet_type = json
+        .get("wallet_type")
+        .ok_or_else(|| "Failed to parse wallet_type".to_string())?
+        .to_string();
+    let wallet_type = get_wallet_type(&wallet_type)?;
+    match wallet_type {
+        WalletType::Standard => {
+            let keystore = json
+                .get("keystore")
+                .ok_or_else(|| "Failed to parse the keystore for single sig".to_string())?;
+            let exkey = get_xkey_from_keystore(keystore)?;
+            let desc_ext = exkey.kind().to_string() + "(" + &exkey.xkeystr() + "/0/*)";
+            let desc_chg = exkey.kind().to_string() + "(" + &exkey.xkeystr() + "/1/*)";
+            Ok(vec![desc_ext, desc_chg])
+        }
+        WalletType::Multisig(x, y) => {
+            let mut desc = String::new();
+            for i in 1..(y + 1) {
+                let query = format!("x{}/", i);
+                let keystore = json
+                    .get(query)
+                    .ok_or_else(|| "Failed to parse the keystore for multi sig".to_string())?;
+                let exkey = get_xkey_from_keystore(keystore)?;
+                if desc.is_empty() {
+                    let prefix = match &exkey.kind().to_string() as &str {
+                        "pkh" => "sh",
+                        kind => kind,
+                    }
+                    .to_string();
+                    desc = prefix + &format!("(sortedmulti({}", x);
+                }
+                desc += &(",".to_string() + &exkey.xkeystr() + "/0/*");
+            }
+            desc += "))";
+            let opening = desc.matches('(').count();
+            let closing = desc.matches(')').count();
+            if opening > closing {
+                desc += ")"
+            };
+            let desc_chg = desc.replace("/0/*", "/1/*");
+
+            Ok(vec![desc, desc_chg])
+        }
+    }
+}
+
+pub fn get_xkey_from_keystore(
+    keystore: &serde_json::Value,
+) -> Result<Box<dyn ElectrumExtendedKey>, String> {
+    if let Some(xprv) = keystore.get("xprv") {
+        if let Some(xprv) = xprv.as_str() {
+            let exprv = ElectrumExtendedPrivKey::from_str(xprv)?;
+            return Ok(Box::new(exprv));
+        }
+    }
+
+    let xpub = keystore
+        .get("xpub")
+        .ok_or("Failed to find the xpub")?
+        .as_str()
+        .ok_or("Failed to convert the xpub to str")?;
+    let expub = ElectrumExtendedPubKey::from_str(xpub)?;
+    Ok(Box::new(expub))
+}
+
+pub fn get_json_from_wallet_file(wallet: &Path) -> Result<serde_json::Value, String> {
```

Comment:
can we add a typed struct with the electrum wallet format fields instead of using generic `Value`?

Or did you avoid it because it requires other serde dep? I think if we have serde_json at this point we should have also serde to have typed structs. Maybe we can feature gate this wallet_file feature even if available by default and make serde deps optional

Reply:
I'm not sure if I understand correctly. You would create a rust struct representation of the electrum wallet format? I am not sure how well that would behave with an externally owned dynamic format that can change between wallet types and electrum versions. That's why I only access the json fields that are of interest to me. Iam not too familiar with constructing structs for cases like this. Maybe it works well, I just don't know.
I chose serde_json::Value because it has more meaning than a bare String or Vec<u8>. And I separated the two public functions because I wanted to be able to feed a wallet xml that doesn't come from a file.

I also thought about a feature gate for this functionality to have the dependencies optional.

Reply:
> I'm not sure if I understand correctly. You would create a rust struct representation of the electrum wallet format?

Yes, just to be clear, I am referring to the serde feature of mapping rust struct to json, something like:
```
struct ElectrumWalletFile {
 pub wallet_type: Strig,
 pub keystore....
}
```
and the using:
```
let wallet: ElectrumWalletFile = serde_json::from_str(&file_content)?;
//or
let wallet: ElectrumWalletFile = serde_json::from_reader(BufReader::new(file))?;
```

> That's why I only access the json fields that are of interest to me

There are serde flags to not error on unparsed fields if necessary

> I am not sure how well that would behave with an externally owned dynamic format that can change between wallet types and electrum versions.

If future electrum versions break the parsing I probably prefer it over working on unknown format

Reply:
That worked like a charm. The only thing I am not so happy with, are the multisig keystores. Do you know a better solution for that?

Reply:
Looks like it could work with something like this:
https://serde.rs/deserialize-struct.html
But that seems to be very verbose.

Reply:
Implementing custom serialization and de-serialization functions worked. The internal representation and the produced json format look good now. 
But all that explicit ignore stuff in de-serialization is too brittle. It works for the empty test wallets, but not for real world wallets, because I oversimplified the types of the fields I want to ignore.

Reply:
serde::de::IgnoredAny is what I needed. Looks good now.